pennylane/ops/qubit.py

```diff
@@ -565,6 +567,176 @@ def decomposition(phi, theta, omega, wires):
         return decomp_ops
 
 
+class MultiRZ(Operation):
```

Comment:
This could potentially be achieved with `qml.broadcast` I assume. But then, one would not enjoy the benefits of making use of the diagonal structure that these operators would have. Would it make sense to add a `pattern=diagonal` option to `qml.broadcast` and generalize the benefits used here? (Definitely out of scope here)

Reply:
I'm not sure this follows too much, as `qml.broadcast` is a function for automatically broadcasting specific gates on multiple wires and building up different gate constructions. However, it is up to the plugin to determine how these gates are _implemented_, and we should not worry about implementation details in PennyLane core.

One thing we could do, however: since default qubit now receives the entire circuit graph, default qubit could potentially 'pre-optimize' the circuit graph. For example, it could [contract multiple `RZ` nodes and CNOTs](https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.minors.contracted_nodes.html) that are all part of one layer into a single `MultiRZ` gate. But this would be quite complicated!

Reply:
First, `MultiRZ` is not the same thing as multiple `RZ` gates. Therefore you can't use `broadcast`. One could probably use `broadcast` in the construction of the decomposition.

One could indeed pre-optimize. And that would be especially helpful if multiple `PauliRot` gates are used after each other. Consider for example `RXZZZ` followed by `RYZZZ`. In this case you're building up some sort of ["arpeggio"](https://www.jazzguitar.be/blog/wp-content/uploads/2017/10/minor-arpeggio-tabs.gif) of CNOT and Z gates interleaved with gates that transform `Z` to `X` or `Y` if necessary. In the case above, a lot of those gates would actually cancel out:
```
 0: ───H──────╭X──RZ(0.1)──╭X───H───RX(1.571)──────────╭X──RZ(0.2)──╭X───RX(-1.571)──────┤ ⟨Z⟩
 1: ──────╭X──╰C───────────╰C──╭X──────────────────╭X──╰C───────────╰C──╭X───────────────┤
 2: ──╭X──╰C───────────────────╰C──╭X──────────╭X──╰C───────────────────╰C───────────╭X──┤
 3: ──╰C───────────────────────────╰C──────────╰C────────────────────────────────────╰C──┤
```
You see that we have 4 superfluous CNOTs here.

Reply:
I think the time complexity of gate fusing and removal would not be too harsh becaues you can just walk the circuit from start to end and then have a look into the "future" in the graph. If you can fuse, you do it, otherwise you take the next step.

Reply:
> I think the time complexity of gate fusing and removal would not be too harsh becaues you can just walk the circuit from start to end and then have a look into the "future" in the graph. If you can fuse, you do it, otherwise you take the next step.

🎹 

And it should be something the user can turn off easily if desired!

Reply:
Totally. On many hardware platforms it's actually recommended to have repeated `XX` on idle qubits so they decohere less.